server.go

```diff
@@ -1435,16 +1441,28 @@ func (s *Server) Serve(ln net.Listener) error {
 //
 // Shutdown does not close keepalive connections so its recommended to set ReadTimeout to something else than 0.
 func (s *Server) Shutdown() error {
-	if s.ln != nil {
-		if err := s.ln.Close(); err != nil {
+	s.mu.Lock()
+	ln := s.ln
+	s.ln = nil
+	s.mu.Unlock()
+	if ln != nil {
+		if err := ln.Close(); err != nil {
 			return err
 		}
 	}
 	atomic.StoreInt32(&s.stop, 1)
 	// Closing the listener will make Serve() call Stop on the worker pool.
 	// Setting .stop to 1 will make serveConn() break out of its loop.
 	// Now we just have to wait until all workers are done.
-	s.wg.Wait()
+	for {
+		if open := atomic.LoadInt32(&s.open); open == 0 {
+			break
+		}
+		// This is not an optimal solution but using a sync.WaitGroup
+		// here causes data races as it's hard to prevent Add() to be called
+		// while Wait() is waiting.
+		time.Sleep(time.Millisecond * 100)
```

Comment:
![image](https://user-images.githubusercontent.com/9026581/45720117-896c5e00-bb92-11e8-8f5d-0c113247a468.png)

please, do support at least x86 0xf3 0x90.

Reply:
@kirillDanshin I don't think you understand what this loop does. It's not a busy loop. Adding a `0xf3 0x90` doesn't make any sense.

Of course `time.Sleep` will probably do `0xf3 0x90` if there is no other goroutine waiting to run.

Since the time it takes `Shutdown` to return isn't critical the extra 100 milliseconds of potential wait time doesn't matter. We could also change it to 1 second but that might be a bit too much.

Reply:
Is it critical to not have a `time.Sleep` in `Shutdown` ?

Reply:
No, I would say it doesn't matter if `Shutdown` takes 1ms or 1000ms or longer. You're going to have to wait for active connections anyways so there is no guarantee of the time `Shutdown` takes.